File: apps/desktop/src/services/NativeService/mcp.ts
// Copyright (c) 2026. 千诚. Licensed under GPL v3.

/**
 * MCP Native Commands
 *
 * 封装所有 MCP 相关的 Rust 后端命令
 */

import { invoke } from '@tauri-apps/api/core';

// ===== Types =====

export type McpTransportType = 'stdio' | 'sse' | 'http';

export interface McpServerConfig {
    id: number;
    name: string;
    transport_type: McpTransportType;
    command?: string;
    args?: string[];
    env?: Record<string, string>;
    cwd?: string;
    url?: string;
    headers?: Record<string, string>;
    enabled: boolean;
    tool_timeout: number;
}

export interface McpToolDefinition {
    name: string;
    description?: string;
    input_schema: Record<string, unknown>;
}

export interface McpToolContent {
    type: 'text' | 'image' | 'resource';
    text?: string;
    data?: string;
    mime_type?: string;
    uri?: string;
    blob?: string;
}

export interface McpToolCallResponse {
    success: boolean;
    content: McpToolContent[];
    is_error: boolean;
}

export type McpServerStatus = 'disconnected' | 'connecting' | 'connected' | 'error';

export interface McpServerStatusInfo {
    server_id: number;
    status: McpServerStatus;
    error?: string;
    version?: string;
}

// ===== Commands =====

/**
 * 连接到 MCP 服务器
 */
export async function connectServer(config: McpServerConfig): Promise<void> {
    return invoke('mcp_connect_server', { config });
}

/**
 * 断开 MCP 服务器连接
 */
export async function disconnectServer(serverId: number): Promise<void> {
    return invoke('mcp_disconnect_server', { serverId });
}

/**
 * 列出服务器的所有工具
 */
export async function listTools(serverId: number): Promise<McpToolDefinition[]> {
    return invoke('mcp_list_tools', { serverId });
}

/**
 * 调用 MCP 工具
 */
export async function callTool(
    serverId: number,
    toolName: string,
    args: Record<string, unknown>
): Promise<McpToolCallResponse> {
    return invoke('mcp_call_tool', {
        serverId,
        toolName,
        arguments: args,
    });
}

/**
 * 获取服务器状态
 */
export async function getServerStatus(serverId: number): Promise<McpServerStatusInfo> {
    return invoke('mcp_get_client_status', { serverId });
}

/**
 * 批量获取所有服务器状态
 */
export async function getAllServerStatuses(): Promise<McpServerStatusInfo[]> {
    return invoke('mcp_get_all_client_statuses');
}

/**
 * 断开所有服务器连接
 */
export async function disconnectAll(): Promise<void> {
    return invoke('mcp_disconnect_all');
}

File: apps/desktop/src-tauri/src/commands/mcp.rs
// Copyright (c) 2026. 千诚. Licensed under GPL v3.

//! MCP 进程管理命令。

use crate::core::mcp::{
    McpClientManager, McpServerConfig, McpToolCallResponse, McpToolDefinition, ServerStatusInfo,
};
use tauri::State;

// ===== MCP 客户端命令 =====

#[tauri::command]
pub async fn mcp_connect_server(
    client_manager: State<'_, McpClientManager>,
    config: McpServerConfig,
) -> Result<(), String> {
    client_manager.connect_server(config).await
}

#[tauri::command]
pub async fn mcp_disconnect_server(
    client_manager: State<'_, McpClientManager>,
    server_id: i64,
) -> Result<(), String> {
    client_manager.disconnect_server(server_id).await
}

#[tauri::command]
pub async fn mcp_list_tools(
    client_manager: State<'_, McpClientManager>,
    server_id: i64,
) -> Result<Vec<McpToolDefinition>, String> {
    client_manager.list_tools(server_id).await
}

#[tauri::command]
pub async fn mcp_call_tool(
    client_manager: State<'_, McpClientManager>,
    server_id: i64,
    tool_name: String,
    arguments: serde_json::Value,
) -> Result<McpToolCallResponse, String> {
    client_manager
        .call_tool(server_id, tool_name, arguments)
        .await
}

#[tauri::command]
pub async fn mcp_get_client_status(
    client_manager: State<'_, McpClientManager>,
    server_id: i64,
) -> Result<ServerStatusInfo, String> {
    client_manager.get_server_status(server_id).await
}

/// 获取全部 MCP 客户端状态。
#[tauri::command]
pub async fn mcp_get_all_client_statuses(
    client_manager: State<'_, McpClientManager>,
) -> Result<Vec<ServerStatusInfo>, String> {
    // 统一走 Result 语义，保持命令层错误处理一致。
    client_manager.get_all_server_statuses_result().await
}

#[tauri::command]
pub async fn mcp_disconnect_all(client_manager: State<'_, McpClientManager>) -> Result<(), String> {
    client_manager.disconnect_all().await
}
